server/tilesource/base.py

```diff
@@ -54,6 +59,104 @@ def __init__(self):
+        if height and not width:
+            width = height * 16
+        if width and not height:
+            height = width * 16
+        if width * sourceHeight > height * sourceWidth:
+            width = max(1, int(sourceWidth * height / sourceHeight))
+        else:
+            height = max(1, int(sourceHeight * width / sourceWidth))
+        return width, height
+
+    def _encodeImage(self, image, encoding='JPEG', jpegQuality=95,
+                     jpegSubsampling=0, **kwargs):
+        """
+        Convert a PIL image into the raw output bytes and a mime type.
+
+        :param image: a PIL image.
+        :param encoding: a valid PIL encoding (typically 'PNG' or 'JPEG').
+                         Must also be in the outputMimeTypes map.
+        :param jpegQuality: the quality to use when encoding a JPEG.
+        :param jpegSubsampling: the subsampling level to use when encoding a
+                                JPEG.
+        """
+        if encoding not in self.outputMimeTypes:
+            raise ValueError('Invalid encoding "%s"' % encoding)
+        if image.width == 0 or image.height == 0:
+            tileData = b''
+        else:
+            output = BytesIO()
+            image.save(output, encoding, quality=jpegQuality,
+                       subsampling=jpegSubsampling)
+            tileData = output.getvalue()
+        return tileData, self.outputMimeTypes[encoding]
+
+    def _getRegionBounds(self, metadata, **kwargs):
+        """
+        Given a set of arguments that can include left, right, top, bottom,
+        sourceWidth, sourceHeight, and units, generate actual pixel values for
+        left, top, right, and bottom.
+
+        :param metadata: the metadata associated with this source.
+        :param **kwargs: optional parameters.  See above.
+        :returns: left, top, right, bottom bounds in pixels.
+        """
+        if kwargs.get('units') not in (None, 'pixel', 'pixels', 'fraction'):
+            raise ValueError('Invalid units "%s"' % kwargs['units'])
+        # Copy kwargs so we don't alter an input dictionary
+        kwargs = kwargs.copy()
+        # Convert fraction units to pixels
+        if kwargs.get('units') == 'fraction':
+            for key in ('left', 'right', 'regionWidth'):
+                if key in kwargs:
+                    kwargs[key] = kwargs[key] * metadata['sizeX']
+            for key in ('top', 'bottom', 'regionHeight'):
+                if key in kwargs:
+                    kwargs[key] = kwargs[key] * metadata['sizeY']
+        # convert negative references to right or bottom offsets
+        for key in ('left', 'right', 'top', 'bottom'):
+            if key in kwargs and kwargs.get(key) < 0:
+                kwargs[key] += metadata[
+                    'sizeX' if key in ('left', 'right') else 'sizeY']
+        # Calculate the region we need to fetch
+        left = kwargs.get(
+            'left', kwargs.get('right') - kwargs.get('regionWidth')
```

Comment:
I have no problem with this logic being inlined, but it's very difficult to mentally parse as it's currently formatted. Can you add some additional whitespace (dropping the 2nd argument down to its own line) and some parentheses (making it obvious that the ternary operator applies to the entire subtraction statement, not just the 2nd term).